torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,401 @@
+            )
+
+        if self.is_registered():
+            raise ValueError(
+                "Cannot chain a parametrization on a parametrization that "
+                "was already registered parametrization."
+            )
+
+        _set_parametrization(self.last_parametrization(), "original", parametrization)
+        return self
+
+    def is_chained(self):
+        r"""Returns True if it is chained to other parametrizations
+        """
+        return is_parametrized(self, "original")
+
+    def is_registered(self):
+        r"""Returns True if it is registered on a module
+        """
+        try:
+            self.original_tensor()
+            return True
+        except ValueError:
+            return False
+
+
+def set_caching(module, tensor_name):
+    r""" Sets up the caching mechanism for a given parametrization. This fuction is
+    automatically invoked when using `register_parametrization` on a parameter or
+    a buffer
+
+    After applying this function, the values of the parametrization will be cached
+    when the contextmanager `torch.nn.cached()` is active
+
+    Args:
+        module (nn.Module): module on which to remove the caching mechanism
+        tensor_name (string): name of the parameter, buffer, or parametrization
+        from which the parametrization will be applied
+    """
+    if not is_parametrized(module, tensor_name):
+        raise ValueError(
+            "The tensor {} in module {} is not parametrized".format(
+                tensor_name, type(module)
+            )
+        )
+
+    global _cache
+    key = (id(module), tensor_name)
+    if key not in _cache:
+        _cache[key] = None
+
+
+def remove_caching(module, tensor_name):
+    r""" Removes a caching mechanism for a given parametrization. This is automatically
+    done when using `remove_parametrization` on a parameter or a buffer
+
+    After applying this function, the values of the parametrization will not be cached even
+    in the presence of the contextmanager `torch.nn.cached()`
+
+    Args:
+        module (nn.Module): module on which to remove the caching mechanism
+        tensor_name (string): name of the parameter, buffer, or parametrization
+        from which the parametrization will be applied
+    """
+    if not is_parametrized(module, tensor_name):
+        raise ValueError(
+            "The tensor {} in module {} is not parametrized".format(
+                tensor_name, type(module)
+            )
+        )
+
+    global _cache
+    key = (id(module), tensor_name)
+    if key in _cache:
+        _cache.pop(key)
+
+
+def has_caching(module, tensor_name):
+    r"""Returns True if module[name] is parametrized and has
+    a caching mechanism
+
+    Args:
+        module (nn.Module): module to query
+        tensor_name (string): attribute in the module to query
+    """
+    global _cache
+    key = (id(module), tensor_name)
```

Comment:
You're using `id(module)` here and `module` below! (By the way please add a test case for that if your existing tests haven't caught this error).

Reply:
Yeah, I did not update the tests because things were changing on a weekly basis. I will write proper thorough tests in a bit, now that we have agreed on the final design